Validation script.

grahamgower opened the pull request:
Supports multiprocessing, because some SLiM simulations will be slow. Running the simulations can be easily done on a multicore server and the stats plotted on a workstation. Now it just needs more stats and simulation scenarios.

Output from `./validation.py -j 8 msprime_basic1 slim_basic1` is attached (which took about 40s to run on my laptop).
[msprime_basic1__slim_basic1.pdf](https://github.com/popsim-consortium/stdpopsim/files/4356655/msprime_basic1__slim_basic1.pdf)

grahamgower commented:
And we have a problem. :(
The attached comparison is between msprime and slim without recapitation. I.e.
```
def msprime_basic1(args):
    cmd = "-e msprime HomSap -c chr1 -l 0.001 100"
    return do_cmd(cmd, *args)

def slim_basic2(args):
    cmd = "-e slim --slim-no-recapitation HomSap -c chr1 -l 0.001 100"
    return do_cmd(cmd, *args)
```

[msprime_basic1__slim_basic2.pdf](https://github.com/popsim-consortium/stdpopsim/files/4358771/msprime_basic1__slim_basic2.pdf)

grahamgower commented:
Times are all out by a factor of 10, which is the default scaling factor. Makes sense I suppose, given times are in units of generations.

andrewkern commented:
okay this is super interesting @grahamgower and exactly what we should try to present in figures in the paper. pinging @jradrion here, as he is thinking about such validation figs. I think the immediate thing is to change the scalars here compare stats as you have already started to.

I vote that we merge this PR and start to trouble shoot the validation. @jeromekelleher that ok?

grahamgower commented:
Thanks for the vote of confidence @andrewkern! I've squashed and rebased to make it less ugly.

grahamgower commented:
Comparisons involving msprime, and slim with no recapitation and no rescaling of time, etc. In the last file, the density of segregating sites is too few to make sensible LD plots.
[onepop_constantN_msprime1__onepop_constantN_slim2.pdf](https://github.com/popsim-consortium/stdpopsim/files/4368026/onepop_constantN_msprime1__onepop_constantN_slim2.pdf)
[twopop_no_migration_msprime1__twopop_no_migration_slim2.pdf](https://github.com/popsim-consortium/stdpopsim/files/4368028/twopop_no_migration_msprime1__twopop_no_migration_slim2.pdf)
[twopop_asymmetric_migration_msprime1__twopop_asymmetric_migration_slim2.pdf](https://github.com/popsim-consortium/stdpopsim/files/4368027/twopop_asymmetric_migration_msprime1__twopop_asymmetric_migration_slim2.pdf)

jeromekelleher commented on validation.py:
```diff
@@ -0,0 +1,492 @@
+    return out_file, t1 - t0
+
+
+"""
```
Slightly odd to use a docstring here, and it might have unintended consequences (I'm not sure what this docstring gets attached to). Best use a comment (banner lines are fine IMO, if you want to split up the file visually)

jeromekelleher commented on validation.py:
```diff
@@ -0,0 +1,492 @@
+    assert "-s" not in cmd and "--seed" not in cmd
+    out_file = os.path.join(out_dir, f"{seed}.trees")
+    full_cmd = cmd + f" -q -o {out_file} -s {seed}".split()
+    t0 = time.time()
```
This is wall time which isn't the right thing I think. We either want to use ``perf_counter`` or to wrap the command we're calling using ``/usr/bin/time`` to do the work for us. See [here](https://github.com/mcveanlab/treeseq-inference/blob/master/src/evaluation.py#L377) for an example of using time.

grahamgower replied:
I'll change it to use perf_counter.

YFI, I got bitten by this (supposedly fixed) GNU time bug last year. https://groups.google.com/forum/#!topic/gnu.utils.help/u1MOsHL4bhg

The fix wasn't included in a release until late 2017 (according to http://ftp.gnu.org/gnu/time/), so there are likely many systems still affected. Unhelpfully, `/usr/bin/time --version` does not report a proper version either, so the only way to check if the bug is present is to allocate a specific amount of memory and see if the /usr/bin/time output is sensible.

jeromekelleher replied:
> YFI, I got bitten by this (supposedly fixed) GNU time bug last year. https://groups.google.com/forum/#!topic/gnu.utils.help/u1MOsHL4bhg
> 
> The fix wasn't included in a release until late 2017 (according to http://ftp.gnu.org/gnu/time/), so there are likely many systems still affected. Unhelpfully, /usr/bin/time --version does not report a proper version either, so the only way to check if the bug is present is to allocate a specific amount of memory and see if the /usr/bin/time output is sensible.

Huh, that is interesting. Looks like a bunch of memory usages I've reported will be off by a factor of 4 then. Oh well.

ps. I looked up [YFI](https://www.internetslang.com/YFI-meaning-definition.asp) to see if this was a new one I hadn't heard of. Interesting slip of the keyboard if it was unintended!

grahamgower replied:
RE: YFI; definitely unintended! :)

jeromekelleher commented on validation.py:
```diff
@@ -0,0 +1,492 @@
+    return stats
+
+
+def linkage_disequilibrium(ts, span=int(1e5), bins=50, min_num_sites=20):
```
``10**5`` rather than ``int(1e5)``?

jeromekelleher commented on validation.py:
```diff
@@ -0,0 +1,492 @@
+def do_simulations(rng, output_folder, num_replicates, num_procs, sim_keys):
+    files = dict()
+    times = dict()
+    with multiprocessing.Pool(num_procs) as pool:
```
Just FYI, the cool kids are using https://docs.python.org/3.5/library/concurrent.futures.html now rather than multiprocessing.

jeromekelleher commented on validation.py:
```diff
@@ -0,0 +1,492 @@
+    files = dict()
+    times = dict()
+    for key in sim_keys:
+        out_dir = os.path.join(output_folder, "trees", key)
```
Pathlib would make this (and the previous) function a bit shorter and easier to read.

jeromekelleher commented on validation.py:
```diff
@@ -0,0 +1,492 @@
+
+
+if __name__ == "__main__":
+    args = parse_args()
```
I think it's better to split the job up explicitly into "run" and "plot" (or some other better names). The running bit will end up taking a lot of time, and you want to be able to do the two things independently. It's especially useful to be able to run plots for simulations that are ongoing (so it's good to always flush out any results that have been calculated when you can).

grahamgower commented:
I've made lots more changes, mostly related to your review, @jeromekelleher. I changed to concurrent.futures, and have interleaved plotting with simulating, so plots are produced when the data is ready. It still uses `map`, rather than submit/futures, so there's probably room for more efficiency by avoiding the current pool synchronisation due to `list(executor.map(...))` each time in the loop. I'm not inclined to try and fix that though.

I dropped the population size for the constantN model comparisons, from human values to 1000. This makes slim without recapitation and without scaling finish in a reasonable amount of time (30 seconds, compared to 6000 seconds before).

Once again, plots are attached. I'm considering generating a simple report that can be uploaded to the wiki. Such a report would include docstrings for the models and stats being plotted as a figure caption.
[onepop_constantN_msprime1__onepop_constantN_slim1.pdf](https://github.com/popsim-consortium/stdpopsim/files/4369629/onepop_constantN_msprime1__onepop_constantN_slim1.pdf)
[onepop_constantN_msprime1__onepop_constantN_slim2.pdf](https://github.com/popsim-consortium/stdpopsim/files/4369630/onepop_constantN_msprime1__onepop_constantN_slim2.pdf)
[onepop_constantN_msprime1__onepop_constantN_slim3.pdf](https://github.com/popsim-consortium/stdpopsim/files/4369631/onepop_constantN_msprime1__onepop_constantN_slim3.pdf)
[twopop_asymmetric_migration_msprime1__twopop_asymmetric_migration_slim1.pdf](https://github.com/popsim-consortium/stdpopsim/files/4369633/twopop_asymmetric_migration_msprime1__twopop_asymmetric_migration_slim1.pdf)
[twopop_asymmetric_migration_msprime1__twopop_asymmetric_migration_slim2.pdf](https://github.com/popsim-consortium/stdpopsim/files/4369634/twopop_asymmetric_migration_msprime1__twopop_asymmetric_migration_slim2.pdf)
[twopop_no_migration_msprime1__twopop_no_migration_slim1.pdf](https://github.com/popsim-consortium/stdpopsim/files/4369635/twopop_no_migration_msprime1__twopop_no_migration_slim1.pdf)
[twopop_no_migration_msprime1__twopop_no_migration_slim2.pdf](https://github.com/popsim-consortium/stdpopsim/files/4369636/twopop_no_migration_msprime1__twopop_no_migration_slim2.pdf)

jeromekelleher commented:
Looks great @grahamgower, shall we merge?

grahamgower commented:
Yes please!

jeromekelleher commented:
Awesome, can you squash the commits please?

grahamgower commented:
Yep, done.

andrewkern commented:
awesome
